[ao] Implemented Outlier Detection Report Generation

vspenubarthi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #81252
* #81251
* #81250
* #81249
* #81174
* __->__ #80937

Summary: This adds the implementation for the report generation for the
Outlier Detector class. This includes both the generation of a
dictionary containing each module that had an observer attached and any
relavent stats collected by the observer that can help shed light on
outlier relavent data or computed metrics. It also includes a string
denoting specific modules that had outliers and gives a bit of insight
into what channels they are contained in.

This contains both the implementation for the report generation for the
outlier detector as well as a test class to test the report generation
functionality.

Test Plan: python test/test_quantization.py TestFxDetectOutliers

Reviewers:

Subscribers:

Tasks:

Tags:

vspenubarthi committed:
[ao] Implemented Outlier Detection Report Generation

Summary: This adds the implementation for the report generation for the
Outlier Detector class. This includes both the generation of a
dictionary containing each module that had an observer attached and any
relavent stats collected by the observer that can help shed light on
outlier relavent data or computed metrics. It also includes a string
denoting specific modules that had outliers and gives a bit of insight
into what channels they are contained in.

This contains both the implementation for the report generation for the
outlier detector as well as a test class to test the report generation
functionality.

Test Plan: python test/test_quantization.py TestFxDetectOutliers

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi committed:
Update on "[ao] Implemented Outlier Detection Report Generation"


Summary: This adds the implementation for the report generation for the
Outlier Detector class. This includes both the generation of a
dictionary containing each module that had an observer attached and any
relavent stats collected by the observer that can help shed light on
outlier relavent data or computed metrics. It also includes a string
denoting specific modules that had outliers and gives a bit of insight
into what channels they are contained in.

This contains both the implementation for the report generation for the
outlier detector as well as a test class to test the report generation
functionality.

Test Plan: python test/test_quantization.py TestFxDetectOutliers

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi committed:
Update on "[ao] Implemented Outlier Detection Report Generation"


Summary: This adds the implementation for the report generation for the
Outlier Detector class. This includes both the generation of a
dictionary containing each module that had an observer attached and any
relavent stats collected by the observer that can help shed light on
outlier relavent data or computed metrics. It also includes a string
denoting specific modules that had outliers and gives a bit of insight
into what channels they are contained in.

This contains both the implementation for the report generation for the
outlier detector as well as a test class to test the report generation
functionality.

Test Plan: python test/test_quantization.py TestFxDetectOutliers

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi committed:
Update on "[ao] Implemented Outlier Detection Report Generation"


Summary: This adds the implementation for the report generation for the
Outlier Detector class. This includes both the generation of a
dictionary containing each module that had an observer attached and any
relavent stats collected by the observer that can help shed light on
outlier relavent data or computed metrics. It also includes a string
denoting specific modules that had outliers and gives a bit of insight
into what channels they are contained in.

This contains both the implementation for the report generation for the
outlier detector as well as a test class to test the report generation
functionality.

Test Plan: python test/test_quantization.py TestFxDetectOutliers

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi committed:
Update on "[ao] Implemented Outlier Detection Report Generation"


Summary: This adds the implementation for the report generation for the
Outlier Detector class. This includes both the generation of a
dictionary containing each module that had an observer attached and any
relavent stats collected by the observer that can help shed light on
outlier relavent data or computed metrics. It also includes a string
denoting specific modules that had outliers and gives a bit of insight
into what channels they are contained in.

This contains both the implementation for the report generation for the
outlier detector as well as a test class to test the report generation
functionality.

Test Plan: python test/test_quantization.py TestFxDetectOutliers

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi committed:
Update on "[ao] Implemented Outlier Detection Report Generation"


Summary: This adds the implementation for the report generation for the
Outlier Detector class. This includes both the generation of a
dictionary containing each module that had an observer attached and any
relavent stats collected by the observer that can help shed light on
outlier relavent data or computed metrics. It also includes a string
denoting specific modules that had outliers and gives a bit of insight
into what channels they are contained in.

This contains both the implementation for the report generation for the
outlier detector as well as a test class to test the report generation
functionality.

Test Plan: python test/test_quantization.py TestFxDetectOutliers

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles commented on torch/ao/quantization/fx/_model_report/detector.py:
```diff
@@ -1038,8 +1150,59 @@ def generate_detector_report(self, model: GraphModule) -> Tuple[str, Dict[str, A
             Dictionary mapping modules of interest to:
                 whether there were outliers found in activation before
+                the number of batches used for each channel
+                whether the number of applicable batches is above statistical threshold (30)
```
would maybe change language because its a kind of arbitrary threshold, we probably don't want to imply statistical significance without knowing anything about relevant distributions.

vspenubarthi replied:
updated to just say above a minimum set threshold so not misleading

HDCharles commented on torch/ao/quantization/fx/_model_report/detector.py:
```diff
@@ -933,6 +933,7 @@ class OutlierDetector(DetectorBase):
 
     Args:
         ratio_threshold (float, optional): The threshold for p_r to determine if there are outliers in activations
+            Should be >= 1
```
would be good to add that statistical threshhold as a parameter here potentially in this PR or a later one

vspenubarthi replied:
will add in a future PR

HDCharles commented on torch/ao/quantization/fx/_model_report/detector.py:
```diff
@@ -1025,6 +1051,92 @@ def determine_observer_insert_points(self, prepared_fx_model: GraphModule) -> Di
+                # also get the max values
+                max_vals: torch.Tensor = pre_obs.max_val
+
+                # we have to specifically modify how we are recording negative ratio for pre-relu layers
```
not sure I follow the logic here and its unclear why "we have to specifically modify how we are recording negative ratio for pre-relu layers"

a counter example or something to illustrate the issue would be helpful.

HDCharles replied:
I think i understand what this is doing, would be good to do this with if statements, its a bit less elegant but its much easier to understand.

vspenubarthi replied:
done with for loop + if statements, not as efficient but def more readable

HDCharles commented on torch/ao/quantization/fx/_model_report/detector.py:
```diff
@@ -1025,6 +1051,92 @@ def determine_observer_insert_points(self, prepared_fx_model: GraphModule) -> Di
+
+        Returns a dictionary mapping:
+            "outliers_detected" : list of bools per channel that are true if it is considered an outlier
+            "statistically_significant": if the per channel calculation had at least 30 samples
```
again the language is a bit suspect

vspenubarthi replied:
changed to say 

`"above_sample_threshold_count": if the per channel calculation had at least 30 samples (a random threshold)`

vspenubarthi replied:
will update when i do the threshold calculation and make it a parameter, so this will be updated soon in a coming PR

vspenubarthi committed:
Update on "[ao] Implemented Outlier Detection Report Generation"


Summary: This adds the implementation for the report generation for the
Outlier Detector class. This includes both the generation of a
dictionary containing each module that had an observer attached and any
relavent stats collected by the observer that can help shed light on
outlier relavent data or computed metrics. It also includes a string
denoting specific modules that had outliers and gives a bit of insight
into what channels they are contained in.

This contains both the implementation for the report generation for the
outlier detector as well as a test class to test the report generation
functionality.

Test Plan: python test/test_quantization.py TestFxDetectOutliers

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi committed:
Update on "[ao] Implemented Outlier Detection Report Generation"


Summary: This adds the implementation for the report generation for the
Outlier Detector class. This includes both the generation of a
dictionary containing each module that had an observer attached and any
relavent stats collected by the observer that can help shed light on
outlier relavent data or computed metrics. It also includes a string
denoting specific modules that had outliers and gives a bit of insight
into what channels they are contained in.

This contains both the implementation for the report generation for the
outlier detector as well as a test class to test the report generation
functionality.

Test Plan: python test/test_quantization.py TestFxDetectOutliers

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi committed:
Update on "[ao] Implemented Outlier Detection Report Generation"


Summary: This adds the implementation for the report generation for the
Outlier Detector class. This includes both the generation of a
dictionary containing each module that had an observer attached and any
relavent stats collected by the observer that can help shed light on
outlier relavent data or computed metrics. It also includes a string
denoting specific modules that had outliers and gives a bit of insight
into what channels they are contained in.

This contains both the implementation for the report generation for the
outlier detector as well as a test class to test the report generation
functionality.

Test Plan: python test/test_quantization.py TestFxDetectOutliers

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi committed:
Update on "[ao] Implemented Outlier Detection Report Generation"


Summary: This adds the implementation for the report generation for the
Outlier Detector class. This includes both the generation of a
dictionary containing each module that had an observer attached and any
relavent stats collected by the observer that can help shed light on
outlier relavent data or computed metrics. It also includes a string
denoting specific modules that had outliers and gives a bit of insight
into what channels they are contained in.

This contains both the implementation for the report generation for the
outlier detector as well as a test class to test the report generation
functionality.

Test Plan: python test/test_quantization.py TestFxDetectOutliers

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

andrewor14 commented on torch/ao/quantization/fx/_model_report/detector.py:
```diff
@@ -1025,6 +1051,97 @@ def determine_observer_insert_points(self, prepared_fx_model: GraphModule) -> Di
+        num_batches_list: List = counted_batches.tolist()
+
+        # calculate whether channels were statistically significant
+        significant_size = [True if batch_size >= 30 else False for batch_size in num_batches_list]
```
Should make this a constant in the class or something instead of hardcoding it

andrewor14 commented on torch/ao/quantization/fx/_model_report/detector.py:
```diff
@@ -1025,6 +1051,97 @@ def determine_observer_insert_points(self, prepared_fx_model: GraphModule) -> Di
+
+        Returns a dictionary mapping:
+            "outliers_detected" : list of bools per channel that are true if it is considered an outlier
+            "above_sample_threshold_count": if the per channel calculation had at least 30 samples (a random threshold)
```
This and other comments should refer to the constant once you add it

andrewor14 commented on torch/ao/quantization/fx/_model_report/detector.py:
```diff
@@ -1025,6 +1051,97 @@ def determine_observer_insert_points(self, prepared_fx_model: GraphModule) -> Di
+        outlier_dict[self.SUFFICIENT_BATCHES_KEY] = significant_size
+
+        # calculate for each channel whether it's an outlier or not based on ratio
+        outlier_detected = [True if ratio > self.ratio_threshold else False for ratio in ratios_list]
```
Can just do `[ratio > self.ratio_threshold for ratio in ratios_list]`

andrewor14 commented on torch/ao/quantization/fx/_model_report/detector.py:
```diff
@@ -1038,8 +1155,59 @@ def generate_detector_report(self, model: GraphModule) -> Tuple[str, Dict[str, A
+            note_composed = note_string.format(note_distribution, note_rec)
+            outlier_string += note_composed
+        else:
+            outlier_string += "There were no outliers found in the activations.\n"
```
This is more of a general suggestion but I still think it'll be useful in a future PR to test the strings returned. There's a lot of string concatenation logic here and it's possible that we end up returning gibberish or malformed sentences to the user and there's no way to catch that currently. This applies to all the detectors you added.

vspenubarthi commented on torch/ao/quantization/fx/_model_report/detector.py:
```diff
@@ -1025,6 +1051,97 @@ def determine_observer_insert_points(self, prepared_fx_model: GraphModule) -> Di
+                # also get the max values
+                max_vals: torch.Tensor = pre_obs.max_val
+
+                # we have to specifically modify how we are recording negative ratio for pre-relu layers
```
done

vspenubarthi committed:
Update on "[ao] Implemented Outlier Detection Report Generation"


Summary: This adds the implementation for the report generation for the
Outlier Detector class. This includes both the generation of a
dictionary containing each module that had an observer attached and any
relavent stats collected by the observer that can help shed light on
outlier relavent data or computed metrics. It also includes a string
denoting specific modules that had outliers and gives a bit of insight
into what channels they are contained in.

This contains both the implementation for the report generation for the
outlier detector as well as a test class to test the report generation
functionality.

Test Plan: python test/test_quantization.py TestFxDetectOutliers

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kit1980 commented on torch/ao/quantization/fx/_model_report/detector.py:
```diff
@@ -983,6 +994,21 @@ def _supports_insertion(self, module: nn.Module) -> bool:
         # check if the module has any children and isn't observer
         num_children = len(list(module.children()))
         return num_children == 0 and not is_activation_post_process(module)
+        # check to see if module is of a supported type
+        num_children = len(list(module.children()))
+        is_supported_type = num_children == 0 and not isinstance(module, ObserverBase)
```
@vspenubarthi Some merge-related issues? The code lines 997-999 is after `return` and will never be executed.

vspenubarthi replied:
@kit1980 seems like it, i'm surprised neither the linter nor tests found this. I created a new PR with the quick fix here: https://github.com/pytorch/pytorch/pull/81499

It would be great if you could give it a quick approval if you think it's all good.

Thank you for finding this.
